RZ AD INS

lindsayad opened the pull request:
@jiangwen84 I'm going to add a coupled of stabilized tests, but hopefully this is enough to get you started

lindsayad committed:
Add RZ terms to AD INS implementation

Closes #14901

lindsayad committed:
Allow not setting components of vector variable dirichlet BCs

This is very useful for example in Navier-Stokes simulations when
using a vector variable implementation in case you want to set
a DirichletBC for one component of the velocity but not another

jiangwen84 commented:
@lindsayad  besides the RZ NS, the change to nodal bc is quite useful too! thanks. Looks like you add support for displaced mesh, for the ALE?

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/507688/) on ba82224 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/15124/site)

This comment will be updated on new commits.

lindsayad commented:
The displaced mesh support was already there, but because I wrote my own `computeResidual` and `computeJacobian` to incorporate the additional RZ term you see the `_use_displaced_mesh` flag showing up

jiangwen84 commented on modules/navier_stokes/src/materials/INSADMaterial.C:
```diff
@@ -66,24 +64,70 @@ INSADMaterial::INSADMaterial(const InputParameters & parameters)
+  // _viscous_strong_residual[_qp] = -_mu[_qp] * Moose::Laplacian(_second_velocity[_qp]);
+
+  if (_coord_sys == Moose::COORD_RZ)
+  {
```
@lindsayad Could you confirm with me: for RZ, _viscous_strong_residual is only scalar part, but not vector part, for Cartesian, it is basically zero. So it means if we use stabilization, it causes small error for bi-linear elements.

lindsayad replied:
There is a piece missing for both Cartesian and cylindrical coordinates and that is the "Cartesian" Laplacian, e.g any second derivative terms.

I may have to update that comment...I seem to recall a long while ago testing the Cartesian AD stabilization method and discovering that for bilinears the error was zero, and that you would only incur error if using biquadratics. I believe this is because the Laplacian includes only the xx, yy, zz derivatives which are zero for bilinears (but wouldn't be zero for biquadratics)

jiangwen84 replied:
It could be no error for bilinear case. I am curious how do you check if there is any error that is coming the stabilization Lap term.

jiangwen84 commented:
> The displaced mesh support was already there, but because I wrote my own `computeResidual` and `computeJacobian` to incorporate the additional RZ term you see the `_use_displaced_mesh` flag showing up

Do we really have to write your own `computeResidual` ? Add RZ term to `computeQpResidual` would not work?

lindsayad committed:
Detailed equivalence testing of hand-coded and AD impls

Refs #14901

lindsayad committed:
Add INSADMomentumNoBCBC

lindsayad commented:
> Do we really have to write your own `computeResidual` ? Add RZ term to `computeQpResidual` would not work?

You could do it that way and the code would definitely be easier to read. I set it up the way it currently is for optimization purposes. There is no need to compute the non-test parts of the residual for every `_i`. By overriding `computeResidual/Jacobian` we can compute the non-test parts of the residual before entering the `_i` loop and avoid a bunch of unnecessary computation which can quickly mount when using vectors and tensors

jiangwen84 commented:
> You could do it that way and the code would definitely be easier to read. I set it up the way it currently is for optimization purposes. There is no need to compute the non-test parts of the residual for every `_i`. By overriding `computeResidual/Jacobian` we can compute the non-test parts of the residual before entering the `_i` loop and avoid a bunch of unnecessary computation which can quickly mount when using vectors and tensors

That is fine. I just want to know if there is any reason other than efficiency that needs to do that way. Thanks.

lindsayad committed:
Add PSPG and SUPG stabilization tests

Verifies that hand-coded and AD functions are equivalent for
a first order basis for the velocity variable

Refs #14901

lindsayad committed:
Add second order stabilization tests (#14901)

lindsayad committed:
Add Jacobian test (#14901)

lindsayad commented:
I've created https://github.com/libMesh/libmesh/issues/2521 for developing second derivative support for vector variables in libMesh

lindsayad committed:
Add INSADMomentumNoBCBC documentation

lindsayad committed:
Add link to John's RZ derivation (#14901)
